### PR TITLE
[pytree] make `context` and `children_specs` as private implementation details

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114947
* __->__ #116375
* #116372
* #116374

This PR makes `treespec.context` and `treespec.children_specs` as private implementation details. It adds property accessors with warnings to keep backward-compatibility.

cc @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @peterbell10 @aakhundov

### CODE REVIEW

File: torch/utils/_pytree.py [hunk 1]
```diff
@@ -482,47 +482,66 @@ def _is_leaf(tree: PyTree) -> bool:
...
+        )
+        return self._context
+
+    @property
+    def children_specs(self) -> List["TreeSpec"]:
+        warnings.warn(
+            "`treespec.children_specs` is private implementation detail. "
+            "It might be changed in the future. "
+            "Please use `treespec.children()` or `treespec.child(idx)` instead.",
+            stacklevel=2,
```
Comment (Skylion007) [inline torch/utils/_pytree.py:528]: Nice job using the stacklevel

File: torch/utils/_pytree.py [hunk 2]
```diff
@@ -482,47 +482,66 @@ def _is_leaf(tree: PyTree) -> bool:
...
 
     num_nodes: int = dataclasses.field(init=False)
     num_leaves: int = dataclasses.field(init=False)
     num_children: int = dataclasses.field(init=False)
 
     def __post_init__(self) -> None:
-        self.num_nodes = 1 + sum(spec.num_nodes for spec in self.children_specs)
-        self.num_leaves = sum(spec.num_leaves for spec in self.children_specs)
-        self.num_children = len(self.children_specs)
+        self.num_nodes = 1 + sum(spec.num_nodes for spec in self._children_specs)
```
Comment (Skylion007) [inline torch/utils/_pytree.py:493]:
```suggestion
        self.num_nodes = sum(spec.num_nodes for spec in self._children_specs, start=1)

**Review by thiagocrepaldi — COMMENTED** [review]
ONNX-related changes are fine by me, but it would be nice an explanation on why these attributes need to become private

**Comment by zou3519** [comment]
mm this is probably too BC-breaking. What's the motivation for making these fields private?

**Review by zou3519 — COMMENTED** [review]
see comment

**Comment by XuehaiPan** [comment]
> mm this is probably too BC-breaking.

I keep the old APIs as read-only properties with deprecation warnings.

------

> What's the motivation for making these fields private?

This question was asked before. See https://github.com/pytorch/pytorch/pull/116372#issuecomment-1897079081 / https://github.com/pytorch/pytorch/pull/116372#issuecomment-1897806555 / https://github.com/pytorch/pytorch/pull/116372#issuecomment-1912646223.

The original thread:

> > > What's the motivation for this? Also, do you want review on the whole stack?

> > > What's the motivation for this?
> > 
> > @zou3519 This is a prerequisite PR to add this explicit API and hide the implementation details.
> > 
> > * [[pytree] make `context` and `children_specs` as private implementation details #116375](https://github.com/pytorch/pytorch/pull/116375)
> > * [[pytree] traverse `dict` in sorted key ordering #114947](https://github.com/pytorch/pytorch/pull/114947)
> > 
> > The motivation is to provide an explicit API for accessing the keys of a dict spec. Because the implementation details for `context` may change (i.e. in #114947 the context is changed `original_keys -> (sorted_keys, original_copy)`). This new API allows us to prevent torch and downstream packages from relying on `spec.context` to provide better backward compatibility.
> > 
> > > Also, do you want review on the whole stack?
> > 
> > Each PR in the stack requires the previous PR. Except for the last PR, PRs in the stack are easy to review.
> >

> > The motivation is to provide an explicit API for accessing the keys of a dict spec. Because the implementation details for context may change (i.e. in #114947 the context is changed original_keys -> (sorted_keys, original_copy)). This new API allows us to prevent torch and downstream packages from relying on spec.context to provide better backward compatibility.
> 
> It might be too late for this, we are able to serialize TreeSpec and that means that the context is a part of the public contract

**Pull request closed without merging**